### PR TITLE
Feat/invalidate profile cache

### PR DESCRIPTION
**Pull request by arghyaiitb**
Handles invalidation of profile data on updation of dataset.

https://trello.com/c/8wPvoYtt/447-the-cache-should-be-invalidated-when-a-new-file-is-added-to-an-existing-dataset

### CODE REVIEW

**Review by milafrerichs — CHANGES_REQUESTED** [review]
Thank you for your work. 

Could you separate the code style changes from the actual work?
We can have a separate discussion about these changes but they're not part of the work described in the PR. 

That would be great. 
thanks 🙏

File: wazimap_ng/cache.py [hunk 1]
```diff
@@ -65,55 +71,86 @@ def last_modified_point_updated(request, profile_id, profile_category_id=None, t
...
+
+@receiver(post_save, sender=points_models.Category)
 def point_updated_category(sender, instance, **kwargs):
     update_point_cache(instance)
 
+
+@receiver(post_save, sender=profile_models.Indicator)
+def indicator_updated(sender, instance, **kwargs):
+    indicator_obj = instance
+    all_profile_indicators = profile_models.ProfileIndicator.objects.filter(
```
Comment (milafrerichs) [inline wazimap_ng/cache.py:133]: What if we consolidate those lookups into just one lookup using `Q` from django?

https://docs.djangoproject.com/en/3.1/topics/db/queries/#complex-lookups-with-q-objects

There is a nice talk from djangoCon on this topic: https://www.youtube.com/watch?v=p9Bgr5VOISo

File: wazimap_ng/cache.py [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
-from django.views.decorators.cache import cache_page, cache_control
-from django.views.decorators.vary import vary_on_headers
 from django.http import Http404
+from django.views.decorators.cache import cache_control
+from django.views.decorators.vary import vary_on_headers
 
-from .profile.models import ProfileIndicator, ProfileHighlight, IndicatorCategory, IndicatorSubcategory, ProfileKeyMetrics, Profile
-from .profile.services import authentication
-from .points.models import Location, Category, Theme
+from wazimap_ng.points import models as points_models
```
Comment (milafrerichs) [inline wazimap_ng/cache.py:11]: why did you change the import to a named import?

File: wazimap_ng/cache.py [hunk 3]
```diff
@@ -19,16 +19,18 @@
 theme_key = "etag-Theme-profile-%s-%s"
 location_theme_key = "etag-Location-Theme-%s"
 
+
```
Comment (milafrerichs) [inline wazimap_ng/cache.py:22]: can you remove all those empty lines please?

**Review by milafrerichs — CHANGES_REQUESTED** [review]
thanks for removing those unrelated changes.

File: wazimap_ng/cache.py (hunk 3 above)
Comment (arghyaiitb) [reply wazimap_ng/cache.py:22]: https://www.python.org/dev/peps/pep-0008/#blank-lines

Can we follow the pep8 convention to add extra line between functions

File: wazimap_ng/cache.py (hunk 2 above)
Comment (arghyaiitb) [reply wazimap_ng/cache.py:11]: Had a quick discussion with @adieyal on imports, we were thinking named imports would have a better readability vs doing too many imports. Happy to take up a discussion on this front.

File: wazimap_ng/cache.py (hunk 1 above)
Comment (arghyaiitb) [reply wazimap_ng/cache.py:133]: Doing it this way the code is much more readable. 
And we are having 4 different tables so we have to use the reverse related query which is sometimes daunting. If you insist I can give it a try but definitely it won't be a clean way to do it. 
Generally I follow this principle 
Code readability > optimisation (unless we have a bottleneck)

File: wazimap_ng/cache.py [hunk 4]
```diff
@@ -65,55 +71,86 @@ def last_modified_point_updated(request, profile_id, profile_category_id=None, t
...
-@receiver(post_save, sender=Category)
+
+@receiver(post_save, sender=points_models.Category)
 def point_updated_category(sender, instance, **kwargs):
     update_point_cache(instance)
 
+
+@receiver(post_save, sender=profile_models.Indicator)
+def indicator_updated(sender, instance, **kwargs):
+    indicator_obj = instance
```
[the remark this answers is not in the document]
Comment (arghyaiitb) [reply wazimap_ng/cache.py]: updated

File: wazimap_ng/cache.py (hunk 1 above)
Comment (milafrerichs) [reply wazimap_ng/cache.py:133]: I agree with your readability should be valued more than optimization. but not sure this applies here. 

```
from django.db.models import Q

profiles_to_invalidate_cache = Profile.objects.fliter(
  Q(indicators__indicator_id=id) 
  | Q(profilekeymetric_set__variable_id=id) 
  | Q(profilehighlight_set__indicator_id=id)
)
```

I don't this is less readable but we can have this discussion at a later point. 
I don't want to halt the merging of this too much.

File: wazimap_ng/cache.py [hunk 5]
```diff
@@ -65,55 +71,86 @@ def last_modified_point_updated(request, profile_id, profile_category_id=None, t
...
+        indicator_id=indicator_id
+    ).values_list('profile_id', flat=True)
+    all_profile_key_metrics = profile_models.ProfileKeyMetrics.objects.filter(
+        variable_id=indicator_id
+    ).values_list('profile_id', flat=True)
+    all_profile_highlights = profile_models.ProfileHighlight.objects.filter(
+        indicator_id=indicator_id
+    ).values_list('profile_id', flat=True)
+    # find the unique profiles whose cache needs to be invalidated
+    list_of_profile_cache_to_be_update = set(all_profile_indicators + all_profile_key_metrics + all_profile_highlights)
```
Comment (milafrerichs) [inline wazimap_ng/cache.py:143]: what about renaming the function name to a more precise name and rename this to `unique_profile_ids_to_invalidate`

**Review by milafrerichs — CHANGES_REQUESTED** [review]
if you can change the imports and decide if you want to change the function and variable name we're good to go.

File: tests/profile/factoryboy.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import factory
```
Comment (milafrerichs) [inline tests/profile/factoryboy.py:1]: I like the addition of factoryboy but this is not used in this PR, right?
could you please remove it? 
thanks.

**Review by milafrerichs — CHANGES_REQUESTED** [review]
almost there :)

File: wazimap_ng/cache.py [hunk 7]
```diff
@@ -65,55 +72,86 @@ def last_modified_point_updated(request, profile_id, profile_category_id=None, t
...
+        indicator_id=indicator_id
+    ).values_list('profile_id', flat=True)
+    all_profile_key_metrics = ProfileKeyMetrics.objects.filter(
+        variable_id=indicator_id
+    ).values_list('profile_id', flat=True)
+    all_profile_highlights = ProfileHighlight.objects.filter(
+        indicator_id=indicator_id
+    ).values_list('profile_id', flat=True)
+    # find the unique profiles whose cache needs to be invalidated
+    unique_profile_ids_to_invalidate = set(list(all_profile_indicators) + list(all_profile_key_metrics) + list(all_profile_highlights))
```
Comment (milafrerichs) [inline wazimap_ng/cache.py:144]: so it seems the test has prevented a bug :) which is awesome :) 
we could use a union here, right? might be cleaner than the set and list calls. 
what do you think? 
`unique_profile_ids_to_invalidate = all_profile_indicators.union(all_profile_key_metrics, all_profile_highlights)`

File: tests/test_cache.py [hunk 8]
```diff
@@ -114,66 +125,21 @@ def test_update_point_cache_signal(mock_datetime):
...
-# def test_point_updated_category(mock_update_point_cache):
-#     category = "Some category"
-
-#     cache.point_updated_category("sender", category)
 
-#     assert mock_update_point_cache.assert_called_with("Some category")
+@pytest.mark.django_db
+class TestCache(unittest.TestCase):
+    @patch('wazimap_ng.cache.update_profile_cache', autospec=True)
+    def test_indicator_updated(self, mock_update_profile_cache):
```
Comment (milafrerichs) [inline tests/test_cache.py:132]: the test name is not very descriptive. what do you think about: `test_invalidate_profile_cache_for_indicator`

and can you add a test where you add more factories with a different indicator to make sure those don't get invalidated?

File: wazimap_ng/cache.py (hunk 7 above)
Comment (adieyal) [reply wazimap_ng/cache.py:144]: Do you prefer `.union` vs the `|` operator?

Comment (milafrerichs) [reply wazimap_ng/cache.py:144]: I would prefer `|` but it does not work here.

```
AssertionError: Cannot combine queries on two different base models.
```

Comment (adieyal) [reply wazimap_ng/cache.py:144]: I was thinking of set unions not queryset unions, but it's minor and probably just a matter of taste

File: wazimap_ng/cache.py [hunk 9]
```diff
@@ -65,55 +73,79 @@ def last_modified_point_updated(request, profile_id, profile_category_id=None, t
...
+
 @receiver(post_save, sender=Category)
 def point_updated_category(sender, instance, **kwargs):
     update_point_cache(instance)
 
+
+@receiver(post_save, sender=Indicator)
+def indicator_updated(sender, instance, **kwargs):
+    indicator_id = instance.id
+    profiles_to_invalidate_cache = Profile.objects.filter(
```
Comment (milafrerichs) [inline wazimap_ng/cache.py:135]: as @arghyaiitb pointed out we should make this distinct. but then we're ready to go.

**Review by milafrerichs — CHANGES_REQUESTED** [review]

**Review by milafrerichs — APPROVED** [review]
